### ISSUE TITLE
Required output field but not input

### ISSUE THREAD

**Issue by benesch** [tfranzel/drf-spectacular#480]
I have a field that is not required on input but will always be present on output. Is it possible to model this field using drf-spectacular? Try as I might, I can't figure out a way to do it, and my Googling is failing me.

I am setting `COMPONENT_SPLIT_REQUEST: True`. So ideally the field would not be marked as required in `FooRequest`, but be marked as required in `Foo`.

Right now I'm setting `required=False`, but that means that a field is not required during deserialization _or_ serialization, which isn't quite right.

**Comment by tfranzel** [comment]
Hi, this is an interesting question. To start off, `COMPONENT_SPLIT_REQUEST: True` would definitely be required for this, as the duality could not be represented otherwise.

On first sight i thought, of course that makes sense. But how would the field actually be written down in the serializer? You can't really do that, right? So DRF is missing the expressive power to describe that circumstance. Then the question would be how to attach that missing information. 

I definitely see the mismatch, but I am unsure how introduce that functionality.

**Comment by benesch** [comment]
> On first sight i thought, of course that makes sense. But how would the field actually be written down in the serializer? You can't really do that, right? So DRF is missing the expressive power to describe that circumstance. Then the question would be how to attach that missing information.

Yep, exactly. Representing this in DRF doesn't seem possible.

**Comment by benesch** [comment]
I thought about this more, and I can't think of a case where I would ever want a field that's optional during serialization. Our API always returns `field: null` in that case, to explicitly indicate the field's absence. So I whipped up this post-processing hook:

```py
def mark_all_outputs_required(result: Any, **kwargs: Any):
    schemas = result.get('components', {}).get('schemas', {})
    for name, schema in schemas.items():
        if name.endswith("Request"):
            continue
        schema["required"] = sorted(schema["properties"].keys())
    return result

```

...and, gosh, that seems to be working great. So, thanks! I guess I just needed confirmation that there wasn't a way to do this in DRF to resort to a post processing hook, and it turned out much simpler than I expected.

Would you like me to close out this issue? One the one hand this definitely isn't a bug in drf-spectacular, but on the other this seems like a super common API pattern, and I'd be delighted if you figured out a better way of supporting it.

**Comment by tfranzel** [comment]
hey, glad that worked out for you!

theoretically we could add a setting to enable this, but i would put that on hold for now until a few more people ask for it. i assume very few people will actually pay attention to this detail. furthermore, i find it difficult to describe why and what this does in one short sentence. so even fewer people will likely understand right away what this setting would do. let's see how this plays out. just leave it open and i'll tag it.

**Comment by expresspotato** [comment]
Correct me if I'm wrong, but we've been doing this successfully with 2 seralizers and settings request / responses as such:

`@extend_schema_view(create=extend_schema(responses=APIAssociationSerializerRead, request=APIAssociationSerializerWrite)`

This also lets us have the read version return nested representations and the write version accept foreign keys or UUID's.

**Comment by benesch** [comment]
Yep, totally, that would work! But `COMPONENT_SPLIT_REQUEST: True` is _really_ nice for us, because it essentially automatically creates the `Read` and `Write` serializers that—I assume—you maintain by hand?

> i find it difficult to describe why and what this does in one short sentence.

Would you agree that it's easier to explain upstream? Like, if a serializer field supported `read_required` and `write_required`, that seems pretty easy to explain? And then drf-spectacular would actually have enough information available to do the right thing.
